New: List with MO's, where products can assinge/consume (Pickinglist)

Humml87 opened the pull request:
View the Issue #23404.

This is a new List, to find all consumable products for manufacturing orders. All columns can be filter and sorting.
![image](https://user-images.githubusercontent.com/78662388/211148807-2bd3999a-c49e-4398-bf1d-3149a94f7758.png)

Dolibarr switch by clicking on a 'Ref.'-Link to the MO->Production (consumeorproduce). The line to consume is prefilled with the warehouse and quantity.
![image](https://user-images.githubusercontent.com/78662388/211149558-fc755a84-6fb6-4e53-8341-ab12f630b9f3.png)

I think this is a WIP-List to use the MRP-Module.
I would be happy and I'm sure, the community have a big benefit in the production workflow.

Best regards
Christian

Commit:
- Create new list.
- Add link in the left navigation.

stickler-ci committed:
Fixing style errors.

eldy commented:
> View the Issue #23404.
> 
> @eldy: is possible to merge in v17, or is this to late?
> 
> Best regards
> Christian

No.
Feature for v17 are freazed since november. Bug fix and only bug fixes are allowed into v17 branch.

eldy commented on htdocs/core/menus/standard/eldy.lib.php:
```diff
@@ -2157,6 +2157,7 @@ function get_left_menu_mrp($mainmenu, &$newmenu, $usemenuhider = 1, $leftmenu =
 			$newmenu->add("", $langs->trans("MenuMRP"), 0, $user->hasRight('mrp',  'read'), '', $mainmenu, 'mo', 0, '', '', '', img_picto('', 'mrp', 'class="paddingright pictofixedwidth"'));
 			$newmenu->add("/mrp/mo_card.php?leftmenu=mo&amp;action=create", $langs->trans("NewMO"), 1, $user->hasRight('mrp',  'write'), '', $mainmenu, 'mo');
 			$newmenu->add("/mrp/mo_list.php?leftmenu=mo", $langs->trans("List"), 1, $user->hasRight('mrp',  'read'), '', $mainmenu, 'mo');
+			$newmenu->add("/mrp/mo_list_assignable.php?leftmenu=mo", $langs->trans("AssignableList"), 1, $user->hasRight('mrp',  'read'), '', $mainmenu, 'mo');
```
I don't understand what is goal of this page.
What means qty missing ? Missing for what ?

Humml87 replied:
The Page views all MO's with the state (validated, in production, produced) where products can consuming.

**Currently:**
After a some stock movement, the warehouse clerk must go on all MO's  MO -> Production then view if its possible to consume a Product.
This is not practicable if you have more then 20 running MO´s. As example in our manufacturing industry (engineering company) we have about 500-1000 MO´s in production at the same time. This is normal in a manufacturing industry.

**With the View:**
After a some stock movement, the warehouse clerk go on the new view and see all MO's where Products can assigne/consume.

"What means qty missing ? Missing for what ?"
=> this means how many quantity of a Product is missing, to reach a complete consuming for a MO-Line.

I hope, i could tell the problem explain well. English is a bit problematical for me ;).

Humml87 replied:
Think you the naming "assign" is bad?
If it's better for the naming in Dolibarr, i can rename all to "consume"?

fappels replied:
To give my thoughts. Maybe do the same like column 'shippable' in customer order by adding column 'producible' in MO list. But do not throw away this new page but make like this PR for customer order: https://github.com/Dolibarr/dolibarr/pull/21596

Humml87 replied:
Hi @fappels,
the shippable in the sales orders work different.
![image](https://user-images.githubusercontent.com/78662388/210597385-71b314e3-15ec-4b34-b839-7a0a615deea3.png)

**Example for the new view:**
- A new Item receipt from a purchase order is booked.
- The clerk looks in the new view and filter the view after the product´s from the item receipt, he find instand the MO's and can consume thir products to the MO.

**Other example:**
- A project need to focus in the company.
- The clerk looks in the new view and filter all MO's where can consume for a defined Project.

P.S. You idea to have a state "complete consumable" is a nice feature for the next step for this PR.
P.P.S. A similar view for sales order is also a good idea, i thought about that too. But in the feature the view's have to much different workflows. Let's think on Services, Workstations and so on.

Commit:
remove dopple extrafields.

Commit:
- naming from 'assign...' to 'consum...'
- Fix extrafields
- Fix product fields

stickler-ci committed:
Fixing style errors.

Commit:
- Fix filter and sorting
- Add Warehouses
- Direct consuming on MO click

stickler-ci committed:
Fixing style errors.

Commit:
fix stickler

Humml87 commented:
Hi @eldy, hi @fappels,
i spended many time to optimize this PR, please try the functionality.
Also the description of the PR is now understandible (i hope ;-) !?
Best regards
Christian

Humml87 commented:
I make a review with following TODO´s:
- [x] Create a GET call for this view in the REST API. This is useful for some PDA-Systems/BDE-Terminals.

Humml87 committed:
Merge branch 'Dolibarr:develop' into develop_#23404_ListMoAssinge

Commit:
Add molineconsumable.php

Commit:
- create molineconsumable.class.php
- create REST API GET -> /mos/consumablemolines
- make mo_list_consumable.php clean (with molineconsumable.class.php)

stickler-ci committed:
Fixing style errors.

Commit:
- fix problems with date filters.
- fix problems to remove filters.

Commit:
add WarehouseToConsume

Humml87 commented:
@eldy I made few pictures from the production. I hope i can show you the benefit of this PR.

We have a Stock with 10.000 of different parts like this
![image](https://user-images.githubusercontent.com/78662388/232684632-b9419061-6881-4e60-8a46-9ff9608e0790.png)

If we create eg. start a new production order, a consuming box is created (one box per production order, sometimes we have 500 boxes on the same time).
![image](https://user-images.githubusercontent.com/78662388/232685251-7844820a-d958-4850-9351-513aa3eb5ac8.png)

Some missing parts for different manufacturing orders comes from a combined purchase order from the suppliers
![image](https://user-images.githubusercontent.com/78662388/232685886-a1002e27-a459-483b-a2da-b287a6f0f91b.png)

The job from the warehouse clerk is to sort the single parts to the manufacturing boxes. For this job it is nessesery to have a overview, witch parts can sort from the warehouse or e.g. from incomming parts can sort to the manufacturing boxes. At a complete consuming the warehouse clerk put a green Paper to the box, so the mechanic pick up the box to produce. Some times the lead time are some weeks.

This PR give the view respectively a joblist for the warehouse clerk.

This is a important feature to use Dolibarr in a machine engineering company.

Humml87 commented:
@eldy wath is the future of this PR?
Please give me feedback. Thank you!

Humml87 commented:
I maked a custom modul for this.

josett225 commented:
Hi @Humml87 
I believe as you this is a good one and mandatory. 
In my side we are building several different bikes sharing most of the parts. 
Preparation of consume share products is key. 
I will test your code if you don't mind. 
Regards 
Jose

Humml87 commented:
> Hi @Humml87 I believe as you this is a good one and mandatory. In my side we are building several different bikes sharing most of the parts. Preparation of consume share products is key. I will test your code if you don't mind. Regards Jose

Hi, feel free. You can find the Code at https://github.com/HUSCH-GmbH/Dolibarr.Custom.MrpBdePicking.
Best regards Christian

josett225 commented:
Hi @Humml87 (Christian)

This is an awesome work you did here!!!
For now it works perfectly as a dream in the first tests we did.
We will update you on next tests in production.

Many many thanks on having share this outstanding custom module.

By the way we are trying to create a new module (Manufacturing Resource Planning) with several BOM when planning our multiple bikes forecast to be able to define our new supplier orders based on MOQ, LeadTime, Stock, sent supplier orders for each product. Let me know if you could be interested in this type of feature.

Best Regards

Jose
